sdk/dynamicconfig/metric/controller/push/period_matcher.go

```diff
@@ -0,0 +1,130 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package push
+
+import (
+	"sync"
+	"time"
+
+	pb "github.com/open-telemetry/opentelemetry-proto/gen/go/experimental/metricconfigservice"
+	"go.opentelemetry.io/contrib/sdk/dynamicconfig/internal/metricpattern"
+	"go.opentelemetry.io/contrib/sdk/dynamicconfig/metric"
+)
+
+const tolerance float64 = 0.1
+
+type PeriodMatcher struct {
+	metrics   map[string]*CollectData
+	startTime time.Time
+
+	m     sync.Mutex
+	sched []*pb.MetricConfigResponse_Schedule
+}
+
+type CollectData struct {
+	lastCollected time.Time
+	period        time.Duration
+}
+
+func (matcher *PeriodMatcher) MarkStart(startTime time.Time) {
+	matcher.startTime = startTime
+}
+
+func (matcher *PeriodMatcher) ConsumeSchedules(sched []*pb.MetricConfigResponse_Schedule) {
+	matcher.m.Lock()
+	defer matcher.m.Unlock()
+
+	matcher.sched = sched
+	matcher.metrics = make(map[string]*CollectData)
+}
+
+// TODO: compute GCD for divisibility issues
```

Comment:
Is this TODO still relevant?

Reply:
Ooops my bad! Left that in accidentally, but already implemented a GCD algorithm for computing the minimum period. The original code intended to use period=SEC_1, in case of indivisibility issues, but I think GCD chooses a period that's a little more efficient.

For the resetting question, the method also initializes the `sched` and `metrics` fields, during the initial gRPC call to obtain schedule data. In this way, I chose `consume` to more generally refer to processing new data, rather than `reset` which has the connotation that the data already exists. Though if `reset` is more broadly intuitive, I can go ahead and change it if you like?